### PR TITLE
Add callback for audio volume range (uac2 only)

### PR DESCRIPTION
**Pull request by electretmike**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new function for retrieving a volume range table, enhancing audio control capabilities.
	- Updated volume control request handling for improved modularity and functionality.

- **Bug Fixes**
	- Adjusted logic for handling request types to ensure correct operations during volume control.

### CODE REVIEW

File: class/audio/usbd_audio.c [hunk 1]
```diff
@@ -190,11 +190,10 @@ static int audio_class_interface_request_handler(uint8_t busid, struct usb_setup
...
                                 if (setup->bmRequestType & USB_REQUEST_DIR_MASK) {
-                                    *((uint16_t *)(*data + 0)) = 1;
-                                    *((uint16_t *)(*data + 2)) = 0;
-                                    *((uint16_t *)(*data + 4)) = 100;
-                                    *((uint16_t *)(*data + 6)) = 1;
-                                    *len = 8;
+                                    usbd_audio_get_volume_table(busid, ep, data);
+                                    num = (uint16_t)((uint16_t)(sampling_freq_table[1] << 8) | ((uint16_t)sampling_freq_table[0]));
+                                    *len = (6 * num + 2);
+                                    memcpy(*data, sampling_freq_table, *len);
```
[the remark this answers is not in the document]
Comment (sakumisu) [reply class/audio/usbd_audio.c:196]: 哪来的 sample freq？

**Comment by sakumisu** [comment]
为什么要加这个API?

**Comment by electretmike** [comment]
> 为什么要加这个API?

I need this, because I use a codec (DAC) that has build-in volume control. I want it to use the proper range. With the current defaults, the host (linux, in my case) will do volume control by changing the values, because the range is way too small. I want to tell the host the supported range of the codec,  so that I keep the dynamic range.

**Comment by sakumisu** [comment]
This range is in uac2.0 spec, so you should convert this range into your dac range.

**Comment by electretmike** [comment]
In the spec, I only find this:

The Volume Control is one of the building blocks of a Feature Unit. A Volume Control must support the
CUR and RANGE(MIN, MAX, RES) attributes. The settings for the CUR, MIN, and MAX attributes can
range from +127.9961 dB (0x7FFF) down to -127.9961 dB (0x8001) in steps of 1/256 dB or 0.00390625
dB (0x0001). The settings for the RES attribute can only have positive values and range from 1/256 dB
(0x0001) to +127.9961 dB (0x7FFF).

I don't see the hardcoded 0..100 range.

**Comment by sakumisu** [comment]
> In the spec, I only find this:
> 
> The Volume Control is one of the building blocks of a Feature Unit. A Volume Control must support the CUR and RANGE(MIN, MAX, RES) attributes. The settings for the CUR, MIN, and MAX attributes can range from +127.9961 dB (0x7FFF) down to -127.9961 dB (0x8001) in steps of 1/256 dB or 0.00390625 dB (0x0001). The settings for the RES attribute can only have positive values and range from 1/256 dB (0x0001) to +127.9961 dB (0x7FFF).
> 
> I don't see the hardcoded 0..100 range.

This is uac1.0  not 2.0, and your api is not common, uca1.0 do not use this.

**Comment by sakumisu** [comment]
You can convert 0-100  to your dac volume, this makes simple for us to use because pc only controls wirh 0-100

**Comment by sakumisu** [comment]
Refer to my latest commit.

**Pull request closed without merging**